From 4efef97f6a2e2d338e3900b25d50d24015d6c32e Mon Sep 17 00:00:00 2001
From: Eric Yang <eric.yang.wiwynn@gmail.com>
Date: Tue, 26 Nov 2024 15:28:02 +0800
Subject: [PATCH 03/25] requester: Validate MCTP EID before removal

Due to potential inaccuracies in MCTP infos obtained from the object
mapper, PLDM may inadvertently remove an existing PLDM device.
To prevent this, add a get property check before removal to ensure the
EID of the device to be removed truly does not exist.

Change-Id: I7586cdca5cd77c3d0a62979ca962745bd0e0970e
Signed-off-by: Eric Yang <eric.yang.wiwynn@gmail.com>
Upstream-Status: Pending
---
 requester/mctp_endpoint_discovery.cpp | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/requester/mctp_endpoint_discovery.cpp b/requester/mctp_endpoint_discovery.cpp
index 614b46e..b73b592 100644
--- a/requester/mctp_endpoint_discovery.cpp
+++ b/requester/mctp_endpoint_discovery.cpp
@@ -264,12 +264,25 @@ void MctpDiscovery::removeFromExistingMctpInfos(MctpInfos& mctpInfos,
 {
     for (const auto& mctpInfo : existingMctpInfos)
     {
+        std::string MCTPObject = std::string(MCTPPath) + "/" +
+                                 std::to_string(std::get<3>(mctpInfo)) + "/" +
+                                 std::to_string(std::get<0>(mctpInfo));
+
         if (std::find(mctpInfos.begin(), mctpInfos.end(), mctpInfo) ==
             mctpInfos.end())
         {
-            removedInfos.emplace_back(mctpInfo);
+            try
+            {
+                pldm::utils::DBusHandler().getDbusPropertyVariant(
+                    MCTPObject.c_str(), "EID", MCTPEndpoint::interface);
+            }
+            catch (const sdbusplus::exception::SdBusError& e)
+            {
+                removedInfos.emplace_back(mctpInfo);
+            }
         }
     }
+
     for (const auto& mctpInfo : removedInfos)
     {
         info("Removing Endpoint networkId '{NETWORK}' and  EID '{EID}'",
-- 
2.53.0-Meta

